### ISSUE TITLE
Symbolic einsum support?

### ISSUE THREAD

**Issue by refraction-ray** [dgasmith/opt_einsum#142]
We can easily define some python duck types with ``__add__ ``, ``__mul__`` and so on, and do computations on them as a symbol. Actually many numpy primitives including ``tensordot`` already support such symbolic computations. However the fallback ``c_einsum`` in numpy doesn't support non numeric types, which may make the symbolic einsum fail if no-blas contraction is involved and ``c_einsum`` is directly invoked.

To be specific, please see this gist as an example on what I mean: https://gist.github.com/refraction-ray/dc22288a0d9e22bb263e59487ad8f5ea 

Is there a way currently in opt_einsum making these symbolic einsum work? If not, I guess the way out is to implement a pure python version of ``c_einsum``? It may be very slow but it anyhow works and can serve as a fallback fn for non numeric types?

**Comment by jcmgray** [comment]
I'm not sure about putting a pure python implementation of einsum into ``opt_einsum`` (really is the opposite of 'opt'!) but here is a workaround for you.
```python
def oinsum(eq, *arrays):
    """A ``einsum`` implementation for ``numpy`` object arrays.
    """
    import numpy as np
    import functools
    import operator

    lhs, output = eq.split('->')
    inputs = lhs.split(',')

    sizes = {}
    for term, array in zip(inputs, arrays):
        for k, d in zip(term, array.shape):
            sizes[k] = d

    out_size = tuple(sizes[k] for k in output)
    out = np.empty(out_size, dtype=object)

    inner = [k for k in sizes if k not in output]
    inner_size = [sizes[k] for k in inner]

    for coo_o in np.ndindex(*out_size):

        coord = dict(zip(output, coo_o))

        def gen_inner_sum():
            for coo_i in np.ndindex(*inner_size):
                coord.update(dict(zip(inner, coo_i)))

                locs = [tuple(coord[k] for k in term) for term in inputs]
                elements = (array[loc] for array, loc in zip(arrays, locs))
                yield functools.reduce(operator.mul, elements)

        out[coo_o] = functools.reduce(operator.add, gen_inner_sum())

    return out
```
and then can override what ``einsum`` ``opt_einsum`` uses for ``numpy``, by reaching behind the scenes slightly...:
```python
from opt_einsum.backends.dispatch import _cached_funcs
_cached_funcs['einsum', 'numpy'] = oinsum
```
That seems to work at least for some basic examples.

**Comment by refraction-ray** [comment]
@jcmgray , thanks for your reply, the code indeed helps. Basically, ``oinsum`` in your code snippet is what I mean by "a pure python impl of einsum". Such a fallback implementation has nothing to do with the efficiency if the involved type are all numerics. ``oinsum`` should only be invoked by dispatch when the input array is inspected as non standard numeric types.

By including such a fallback impl of einsum, opt_einsum only gains the feature that support symbolic einsum while losing nothing, it is still opt! (all numeric types will be contracted by ``np.einsum`` or ``np.tensordot`` as before, actually the latter already supports duck types).

**Comment by dgasmith** [comment]
I find it odd that `tensordot` seems to support this out of the box: https://numpy.org/doc/1.18/reference/generated/numpy.tensordot.html

I'm a bit hesitant to add such an item in general as it opens up a lot of questions about broadcasting and type casting, especially with a fear that we might inadvertently fall back to this implementation accidentally. This is seeming like more of a case where we can supply the path, but leave it to a different backend to evaluate.

**Comment by jcmgray** [comment]
Yes I guess my worry is that in my experience when an object array reaches these functions its the result of a bug *somewhere else* to do with type casting etc. The problem then being that this function now still runs but literally 10,000+ times slower. This slow down is essentially a silent failure.

I think it would be better to be explicit about it. I.e allowing something like:
```python
oe.register_backend('numpy-object', einsum=oinsum, tensordot=np.tensordot)  # probably would default to numpy implementation so tensordot kwarg redundant
oe.contract(eq, *obj_arrays, backend='numpy-object')
```
Would that cover your use cases @refraction-ray?

**Comment by refraction-ray** [comment]
> Would that cover your use cases

Yes, the ``register_backend`` idea looks great to me. And I guess there is no such API currently for opt_einsum? Though it should be an easy task to add either within the package or outside of it.

Besides, even with such API, I believe it is great that "numpy_object" backend with ``oinsum`` is pre registered in opt_einsum which enable ``oe.contract(eq, **object-array, backend="numpy-object")`` working out of the box. Then there is no fear of numeric types wrongly falling back on this, since the backend must be specified explicitly.

**Comment by jcmgray** [comment]
I'd be happy to put in a ``backend='object'`` option. I notice it might resolve #86 too. 

The ``register_backend`` function might not be needed then here but maybe useful to add anyhow.

**Comment by dgasmith** [comment]
Cool, the `object` backend would be great and running it through our full test suite will be pretty exhaustive in terms of cases.

**Comment by dgasmith** [comment]
Closed by #145.
